diff --git a/gdk/win32/gdkevents-win32.c b/gdk/win32/gdkevents-win32.c
index 1cde4fc..f403302 100644
--- a/gdk/win32/gdkevents-win32.c
+++ b/gdk/win32/gdkevents-win32.c
@@ -2796,7 +2796,19 @@ gdk_event_translate (MSG  *msg,
       if (_gdk_display->keyboard_grab.window != NULL &&
 	  !GDK_WINDOW_DESTROYED (_gdk_display->keyboard_grab.window))
 	{
-	  generate_grab_broken_event (_gdk_display->keyboard_grab.window, TRUE, NULL);
+	  GdkPointerGrabInfo *grab;
+
+          if (_gdk_display->keyboard_grab.window)
+	    _gdk_display_unset_has_keyboard_grab (_gdk_display, FALSE);
+
+	  grab = _gdk_display_get_last_pointer_grab (_gdk_display);
+	  if (grab)
+	    {
+	      grab->serial_end = 0;
+	      grab->implicit_ungrab = TRUE;
+	    }
+
+	  _gdk_display_pointer_grab_update (_gdk_display, 0);
 	}
 
       /* fallthrough */
@@ -3467,6 +3479,11 @@ gdk_event_translate (MSG  *msg,
 	  break;
 	}
 
+      if (LOWORD (msg->wParam) == WA_INACTIVE)
+	{
+	  generate_grab_broken_event (_gdk_display->keyboard_grab.window, TRUE, NULL);
+	}
+
       /* Bring any tablet contexts to the top of the overlap order when
        * one of our windows is activated.
        * NOTE: It doesn't seem to work well if it is done in WM_ACTIVATEAPP
